### PR TITLE
feat: implement set_code

### PR DESCRIPTION
**Pull request by sabotagebeats**
### What I did

blocking: # https://github.com/ApeWorX/ape/pull/1090

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [x] New test cases have been added and are passing
- [x] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by sabotagebeats** [comment]
needs tests

File: ape_hardhat/provider.py [hunk 1]
```diff
@@ -320,6 +320,9 @@ def build_command(self) -> List[str]:
     def set_block_gas_limit(self, gas_limit: int) -> bool:
         return self._make_request("evm_setBlockGasLimit", [hex(gas_limit)]) is True
 
+    def set_code(self, code):
```
Comment (unparalleled-js) [inline ape_hardhat/provider.py:323]: does this need a return type? Does this actually return a bool?

Comment (sabotagebeats) [reply ape_hardhat/provider.py]: I have left it as bool for now but we can adjust it as needed once it functions properly.

File: tests/test_fork_provider.py [hunk 2]
```diff
@@ -99,6 +99,16 @@ def test_reset_fork_specify_block_number_via_config(networks, create_fork_provid
     provider.disconnect()
 
 
+@pytest.mark.sync
+def test_set_code(create_fork_provider, fork_contract_instance):
```
Comment (unparalleled-js) [inline tests/test_fork_provider.py:103]: This is not really related to `fork` mode so this test can be put in `test_provider.py` and use the normal provider.

Comment (sabotagebeats) [reply tests/test_fork_provider.py:103]: ok great

File: ape_hardhat/provider.py [hunk 3]
```diff
@@ -320,6 +320,9 @@ def build_command(self) -> List[str]:
     def set_block_gas_limit(self, gas_limit: int) -> bool:
         return self._make_request("evm_setBlockGasLimit", [hex(gas_limit)]) is True
 
+    def set_code(self, address: AddressType, code) -> bool:
+        return self._make_request("hardhat_setCode", [address, code]) is True
```
[the remark this answers is not in the document]
Comment (sabotagebeats) [reply ape_hardhat/provider.py:324]: I'm not sure if this is returning a bool: 

```
ape.exceptions.ProviderError: Errors encountered in param 1: Invalid value ["0x608060405234801561001057600080fd5b506004361061009e5760003560e01c80636cbceeec116100665780636cbceeec146101b857806370a08231146101dc57806395d89b4114610202578063a9059cbb1461020a578063dd62ed3e146102365761009e565b806306fdde03146100a3578063095ea7b31461012057806318160ddd1461016057806323b872dd1461017a578063313ce567146101b0575b600080fd5b6100ab610264565b6040805160208082528351818301528351919283929083019185019080838360005b838110156100e55781810151838201526020016100cd565b50505050905090810190601f1680156101125780820380516001836020036101000a031916815260200191505b509250505060405180910390f35b61014c6004803603604081101561013657600080fd5b506001600160a01b0381351690602001356102f1565b604080519115158252519081900360200190f35b610168610358565b60408051918252519081900360200190f35b61014c6004803603606081101561019057600080fd5b506001600160a01b0381358116916020810135909116906040013561035e565b610168610436565b6101c061043c565b604080516001600160a01b039092168252519081900360200190f35b610168600480360360208110156101f257600080fd5b50356001600160a01b0316610440565b6100ab61045b565b61014c6004803603604081101561022057600080fd5b506001600160a01b0381351690602001356104b6565b6101686004803603604081101561024c57600080fd5b506001600160a01b03813581169160200135166104cc565b60018054604080516020600284861615610100026000190190941693909304601f810184900484028201840190925281815292918301828280156102e95780601f106102be576101008083540402835291602001916102e9565b820191906000526020600020905b8154815290600101906020018083116102cc57829003601f168201915b505050505081565b3360008181526005602090815260408083206001600160a01b038716808552908352818420869055815186815291519394909390927f8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925928290030190a35060015b92915050565b60035481565b6001600160a01b03831660009081526005602090815260408083203384529091528120548211156103cf576040805162461bcd60e51b8152602060048201526016602482015275496e73756666696369656e7420616c6c6f77616e636560501b604482015290519081900360640190fd5b6001600160a01b03841660009081526005602090815260408083203384529091529020546103fd90836104f7565b6001600160a01b038516600090815260056020908152604080832033845290915290205561042c84848461050c565b5060019392505050565b60025481565b3090565b6001600160a01b031660009081526004602052604090205490565b6000805460408051602060026001851615610100026000190190941693909304601f810184900484028201840190925281815292918301828280156102e95780601f106102be576101008083540402835291602001916102e9565b60006104c333848461050c565b50600192915050565b6001600160a01b03918216600090815260056020908152604080832093909416825291909152205490565b60008282111561050657600080fd5b50900390565b6001600160a01b038316600090815260046020526040902054811115610570576040805162461bcd60e51b8152602060048201526014602482015273496e73756666696369656e742062616c616e636560601b604482015290519081900360640190fd5b6001600160a01b03831660009081526004602052604090205461059390826104f7565b6001600160a01b0380851660009081526004602052604080822093909355908416815220546105c2908261061e565b6001600160a01b0380841660008181526004602090815260409182902094909455805185815290519193928716927fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef92918290030190a3505050565b8181018281101561035257600080fdfea2646970667358221220f6e319aa58d727b34aff6abffdcb47a8f4e697706c3fe770edf93d5a1d8c4fc564736f6c634300060c0033"] supplied to : DATA
```

File: tests/test_fork_provider.py (hunk 2 above)
Comment (sabotagebeats) [reply tests/test_fork_provider.py:103]: I moved these but theyre failing now.

Comment (unparalleled-js) [reply tests/test_fork_provider.py:103]: It looks like it was copied and not moved

**Review by unparalleled-js — APPROVED** [review]

File: ape_hardhat/provider.py [hunk 4]
```diff
@@ -320,6 +320,15 @@ def build_command(self) -> List[str]:
     def set_block_gas_limit(self, gas_limit: int) -> bool:
         return self._make_request("evm_setBlockGasLimit", [hex(gas_limit)]) is True
 
+    def set_code(self, address: AddressType, code: Union[str, bytes, HexBytes]) -> bool:
```
Comment (unparalleled-js) [inline ape_hardhat/provider.py:323]: As a future: we may want to add a custom type `ape.types`:

```python
ContractCode = Union[str, bytes, HexBytes]
```
and have that be the return type for `get_code()` and the param type here for code in `set_code()`

Comment (sabotagebeats) [reply ape_hardhat/provider.py:323]: should we make a new PR for this?

**Review by unparalleled-js — APPROVED** [review]

File: ape_hardhat/provider.py (hunk 4 above)
Comment (unparalleled-js) [reply ape_hardhat/provider.py:323]: It will be on the Ape core PR, not this one, this PR is good to go

**Review by unparalleled-js — APPROVED** [review]
Let's do this thing

File: ape_hardhat/provider.py [hunk 5]
```diff
@@ -320,6 +320,15 @@ def build_command(self) -> List[str]:
...
+        if isinstance(code, bytes):
+            code = code.hex()
+
+        elif not is_hex(code):
```
Comment (NotPeopling2day) [inline ape_hardhat/provider.py:327]: Didn't know about `is_hex()`, cool!

**Review by NotPeopling2day — APPROVED** [review]
